src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -246,6 +247,208 @@ def requires_python38_or_newer(self, interpreter_universe: Iterable[str]) -> boo
             allowed_versions=py38_and_later, prior_version="3.7"
         )
 
+    def flatten(self, interpreter_universe: Iterable[str]) -> SpecifierSet:
+        """Flatten into a single specifier set.
+
+        Each element in the original list gets ORed, which we must preserve. We do this by including
+        the whole universe but excluding anything not in the original range.
```

Comment:
I can't help but think that this code would be a lot simpler if we more literally did what you say here. The downside is that it would never output ranges (and is maybe less efficient? but probably not in a meaningful way). To avoid an explosion, could include the entire `minor` version if all known `patch` versions matched.

Then `["==2.7.*", ">=3.5,<3.8"]` would become `"==2.7.*,==3.5.*,==3.6.*,==3.7.*"`.

Reply:
> I'm a bit worried about maintaining this code, but it seems well tested.

Understandable, ack that it's complex. Took a very hot minute to figure out and implement :)

--

> I can't help but think that this code would be a lot simpler if we more literally did what you say here and iterated over the universe to match it. 

That could work well in some cases. But when you don't allow every patch for a major/minor, it would become frustratingly verbose. `['>=3.6.2', '==3.7.*']` would become `'==3.6.2,==3.6.3,==3.6.4,...,==3.6.17,==3.16.18,==3.6.19,==3.7.*'`. That hit in readability doesn't seem worth it to me.

And I found this morning that Poetry's error messages _will_ output what the Python version was set to when deps aren't compatible with it. Meaning, users will end up seeing these flattened constraints.

Reply:
> `['>=3.6.2', '==3.7.*']` would become `'==3.6.2,==3.6.3,==3.6.4,...,==3.6.17,==3.16.18,==3.6.19,==3.7.*'`. That hit in readability doesn't seem worth it to me.

It wouldn't need to: see my comment.:

> To avoid an explosion of ranges, could include the entire `minor` version if all known `patch` versions matched.

Reply:
Yes, but I'm saying when _not_ all patch versions match. Black's constraints need to be >=3.6.2, not >=3.6. Which means the explosion of ranges.

(I mean not technically because Black is only a single constraint, so we don't need to flatten. But point being: it is common enough that you don't allow all patch version for a major.minor. I remember seeing in users' config things like `>=3.7.3` for example)